internal/xfn/function_runner_metrics.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2023 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+this file except in compliance with the License. You may obtain a copy of the
+License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed
+under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+*/
+
+package xfn
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/status"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/fn/proto/v1beta1"
+)
+
+// Metrics are requests, errors, and duration (RED) metrics for composition
+// function runs.
+type Metrics struct {
+	requests  *prometheus.CounterVec
+	responses *prometheus.CounterVec
+	duration  *prometheus.HistogramVec
+}
+
+// NewMetrics creates metrics for composition function runs.
+func NewMetrics() *Metrics {
+	return &Metrics{
+		requests: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "composition",
+			Name:      "run_function_request_total",
+			Help:      "Total number of RunFunctionRequests sent.",
+		}, []string{"name", "package", "target"}),
+
+		responses: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: "composition",
+			Name:      "run_function_response_total",
+			Help:      "Total number of RunFunctionResponses received.",
+		}, []string{"name", "package", "target", "grpc_code", "severity"}),
+
+		duration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Subsystem: "composition",
+			Name:      "run_function_seconds",
+			Help:      "Histogram of RunFunctionResponse latency (seconds).",
+			Buckets:   prometheus.DefBuckets,
+		}, []string{"name", "package", "target", "grpc_code", "severity"}),
```

Comment:
I could be convinced to drop `grpc_code` and `severity` on the histogram - could be too much cardinality.

Reply:
what are name and package and target? No obvious from reading here.

Reply:
Updated to make the label names more descriptive.